core/node/da_clients/src/avail.rs

```diff
@@ -0,0 +1,404 @@
+    types::{DAError, DispatchResponse, InclusionData},
+    DataAvailabilityClient,
+};
+
+const PROTOCOL_VERSION: u8 = 4;
+
+/// An implementation of the `DataAvailabilityClient` trait that interacts with the Avail network.
+#[derive(Debug)]
+pub struct AvailClient {
+    config: AvailConfig,
+    keypair: Keypair,
+}
+
+// Utility type needed for encoding the call data
+#[derive(parity_scale_codec::Encode, scale_encode::EncodeAsType)]
+#[encode_as_type(crate_path = "scale_encode")]
+pub struct SubmitData {
+    pub data: BoundedVec<u8>,
+}
+
+// Utility type needed for encoding the call data
+#[derive(parity_scale_codec::Encode, scale_encode::EncodeAsType)]
+#[encode_as_type(crate_path = "scale_encode")]
+pub struct BoundedVec<_0>(pub Vec<_0>);
+
+#[async_trait]
+impl DataAvailabilityClient for AvailClient {
+    async fn dispatch_blob(
+        &self,
+        _: u32, // batch_number
+        data: Vec<u8>,
+    ) -> anyhow::Result<DispatchResponse, DAError> {
+        let client = WsClientBuilder::default()
+            .build(self.config.api_node_url.as_str())
+            .await
+            .map_err(to_non_retriable_da_error)?;
+        let call_data = self
+            .get_encoded_call(&client, data)
+            .await
+            .map_err(to_non_retriable_da_error)?;
+        let extra_params = self
+            .get_extended_params(&client)
+            .await
+            .map_err(to_non_retriable_da_error)?;
+        let additional_params = self
+            .get_additional_params(&client)
+            .await
+            .map_err(to_non_retriable_da_error)?;
+
+        let signature = self.get_signature(
+            call_data.as_slice(),
+            extra_params.as_slice(),
+            additional_params.as_slice(),
+        );
+
+        let ext = self.get_submittable_extrinsic(
+            signature,
+            extra_params.as_slice(),
+            call_data.as_slice(),
+        );
+        let hex_ext = hex::encode(&ext);
+
+        let block_hash = self
+            .submit_extrinsic(&client, hex_ext.as_str())
+            .await
+            .map_err(to_non_retriable_da_error)?;
+        let tx_id = self
+            .get_tx_id(&client, block_hash.as_str(), hex_ext.as_str())
+            .await
+            .map_err(to_non_retriable_da_error)?;
+
+        Ok(DispatchResponse {
+            blob_id: format!("{}:{}", block_hash, tx_id),
+        })
+    }
+
+    async fn get_inclusion_data(
+        &self,
+        _blob_id: &str,
+    ) -> anyhow::Result<Option<InclusionData>, DAError> {
+        // TODO: implement inclusion data retrieval
+        Ok(Some(InclusionData { data: vec![] }))
+    }
+
+    fn clone_boxed(&self) -> Box<dyn DataAvailabilityClient> {
+        Box::new(AvailClient {
+            config: self.config.clone(),
+            keypair: self.keypair.clone(),
+        })
+    }
+
+    fn blob_size_limit(&self) -> Option<usize> {
+        Some(AvailClient::MAX_BLOB_SIZE)
+    }
+}
+
+impl AvailClient {
```

Comment:
Given that (hopefully) we will use a custom implementation for a while, maybe it'd be better to split the concepts for maintainability:

Create a `pub(super)` module `sdk` in the `avail` module. Add a module doc-comment:
```rust
//! Minimal reimplementation of the Avail SDK client required for the DA client implementation.
//! This is considered to be a temporary solution until a mature SDK is available on crates.io
```

Then, create a structure `RawAvailClient` there and move low-level code and types there.
Finally, store an instance of `RawAvailClient` inside of `AvailClient` and use it as you would use the 3rd party SDK.

You may want to define some helper structures to make interactions between `RawAvailClient` and AvailClient` more intuitive, so that you can write `dispatch_blob` roughly as follows:

```rust
let blob = self.raw_client.encode_blob(..); // Returns a "blob wrapper" type from `raw_client`, not bound to the DA client logic.
let response = self.raw_client.dispatch_blob(..).await; // Returns "raw client specific" response type.
Ok(DispatchResponse::from(response)); // `DispatchResponse` is DA client specific and implements `From` trait
```

This way, the DA client implementation would be less specific and it will be easier to migrate to an alternative implementation in the future.

Reply:
You can look at `ObjectStoreDA` as a hint: imagine that `RawAvailClient` is equivalent to `Arc<dyn ObjectStore>`.